player/command.c

```diff
@@ -7114,9 +7108,13 @@ void mp_option_change_callback(void *ctx, struct m_config_option *co, int flags,
     if (flags & UPDATE_SCREENSAVER)
         update_screensaver_state(mpctx);
 
-    if (flags & UPDATE_VOL)
-        audio_update_volume(mpctx);
-
+    if (flags & UPDATE_VOL) {
+        if (opt_ptr == &opts->ao_volume) {
+            audio_update_ao_volume(mpctx);
+        } else {
+            audio_update_volume(mpctx);
+        }
+    }
```

Comment:
Sorry, I realize I told you to do this earlier, but this isn't needed now that I look at it again. We can just check `if (opt_ptr == &opts->ao_volume)`. The `UPDATE_VOL` flag is completely pointless. You can convert this back and then below the `audio_delay` check put:

```
    if (opt_ptr == &opts->ao_volume) {
        audio_update_ao_volume(mpctx);
```

Reply:
done!
but while doing this I noticed another thing that maybe I should remove in `player/audio.c`
```
@@ -593,8 +607,10 @@ void reinit_audio_chain_src(struct MPContext *mpctx, struct track *track)
     if (recreate_audio_filters(mpctx) < 0)
         goto init_error;

-    if (mpctx->ao)
+    if (mpctx->ao) {
         audio_update_volume(mpctx);
+        audio_update_ao_volume(mpctx);
+    }

     mp_wakeup_core(mpctx);
     return;
```
from what I understood this function is supposed to run if the format changed as explained in the comment,
```
    // If true, there was a format change. output_aformat might have changed,
    // and the implementation drained the filter chain and unset the internal ao
    // reference. The API user needs to call mp_output_chain_set_ao() again.
    // Until this is done, the filter chain will not output new data.
    bool ao_needs_update;
```
since `ao-volume` is set in the system volume it shouldn't lose it's state even if the format changed.
I wasn't able to test this though, I tried to test this by removing the `audio_update_volume(mpctx);` and playing to audio files in different formats but it didn't change the behavior so the format change probably means something else?

Reply:
I think it's probably safer to keep it in here. A reinit could happen for a lot of different reasons and I don't think it should cause any harm. Worse case, you just apply the same volume again.